src/persist-txn/src/operator.rs

```diff
@@ -0,0 +1,427 @@
+                    // it to advance before asking again.
+                    //
+                    // Note that we're asking again with the same input, but
+                    // once the cache is past progress_exclusive (as it will be
+                    // after this update_gt call), we're guaranteed to get an
+                    // answer.
+                    DataListenNext::WaitForTxnsProgress => {
+                        txns_cache.update_gt(&output_progress_exclusive).await;
+                        continue;
+                    }
+                    // The data shard got a write! Loop back above and pass
+                    // through data until we see it.
+                    DataListenNext::ReadDataPast(new_target) => {
+                        read_data_past = new_target;
+                        break;
+                    }
+                    // We know there are no writes in
+                    // `[output_progress_exclusive, new_progress)`, so advance
+                    // our output frontier.
+                    DataListenNext::EmitLogicalProgress(new_progress) => {
+                        assert!(output_progress_exclusive < new_progress);
+                        output_progress_exclusive = new_progress;
+                        cap.downgrade(&output_progress_exclusive);
+                        continue;
+                    }
+                }
+            }
+        }
+    });
+    (passthrough_stream, Rc::new(shutdown_button.press_on_drop()))
+}
+
+/// A helper for subscribing to a data shard using the timely operators.
+///
+/// This could instead be a wrapper around a [Subscribe], but it's only used in
+/// tests and maelstrom, so do it by wrapping the timely operators to get
+/// additional coverage. For the same reason, hardcode the K, V, T, D types.
+///
+/// [Subscribe]: mz_persist_client::read::Subscribe
+pub struct DataSubscribe {
+    worker: Worker<Thread>,
+    data: ProbeHandle<u64>,
+    txns: ProbeHandle<u64>,
+    capture: mpsc::Receiver<EventCore<u64, Vec<(String, u64, i64)>>>,
+    output: Vec<(String, u64, i64)>,
+
+    _tokens: Rc<dyn Any>,
+}
+
+impl std::fmt::Debug for DataSubscribe {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let DataSubscribe {
+            worker: _,
+            data,
+            txns,
+            capture: _,
+            output,
+            _tokens: _,
+        } = self;
+        f.debug_struct("DataSubscribe")
+            .field("data", data)
+            .field("txns", txns)
+            .field("output", output)
+            .finish_non_exhaustive()
+    }
+}
+
+impl DataSubscribe {
+    /// Creates a new [DataSubscribe].
+    pub fn new(
+        name: &str,
+        client: PersistClient,
+        txns_id: ShardId,
+        data_id: ShardId,
+        as_of: u64,
+    ) -> Self {
+        let mut worker = Worker::new(WorkerConfig::default(), Thread::new());
```

Comment:
I was very confused by `Thread::new()` when just reading it in GH. Thought this was creating a thread... :sweat_smile: